# Obbligazioni

## Tasso di interesse

Per calcolare gli interessi delle obbligazioni, si usa la formula dell'interesse semplice:

I=P×r×t

dove:

- I è l'importo degli interessi che si desidera calcolare,
- P è il valore nominale dell'obbligazione (il suo prezzo di emissione),
- r è il tasso di interesse annuo dell'obbligazione espresso come una frazione decimale
- t è il periodo di tempo in anni per cui l'obbligazione è stata detenuta.

Se hai bisogno di calcolare l'interesse maturato su un periodo inferiore all'anno (ad esempio, in mesi), dovrai utilizzare una formula leggermente diversa. In questo caso, la formula diventa:

I=P×r×t/12

dove t è è il numero di mesi per cui l'obbligazione è stata detenuta

In [26]:
def calcola_interessi_obbligazione(P, r, t):

    I = P * r * t
    return I

# Esempio di utilizzo:
valore_nominale = 1000  # Prezzo di emissione dell'obbligazione
tasso_interesse_annuo = 0.05  # Tasso di interesse annuo del 5%
periodo_detenzione_anni = 2  # Periodo di detenzione di 2 anni

interessi_obbligazione = calcola_interessi_obbligazione(valore_nominale, tasso_interesse_annuo, periodo_detenzione_anni)
print("Importo degli interessi dell'obbligazione:", interessi_obbligazione)


Importo degli interessi dell'obbligazione: 100.0


## Prezzo

In generale il prezzo delle obbligazioni è rilevato come valore attuale
$$B_0=d_1C_1+d_2C_2+\cdot\cdot\cdot+d_TC_T=\sum_{t=1}^{T}d_tC_t$$

In [27]:
def prezzo_obbligazione_corrente(flussi_di_cassa, tasso_di_interesse_di_mercato):
    prezzo_corrente = sum([flusso / (1 + tasso_di_interesse_di_mercato) ** (i + 1) for i, flusso in enumerate(flussi_di_cassa)])
    return prezzo_corrente

# Esempio di utilizzo della funzione con i valori forniti:
flussi_di_cassa = [10, 10, 110]  # Flussi di cassa dell'obbligazione (inserire quelli che abbiamo)
tasso_di_interesse_di_mercato = 0.09  # Tasso di interesse del mercato (4.5%)

prezzo_obbligazione = prezzo_obbligazione_corrente(flussi_di_cassa, tasso_di_interesse_di_mercato)
print(f"Il prezzo corrente dell'obbligazione è: {prezzo_obbligazione:.3f}")


Il prezzo corrente dell'obbligazione è: 102.531


Come puoi vedere se i tassi di mercato sono maggiori dei tassi offerti dalla obbligazione essa vale di meno per tassi di mercato di solito si fa riferimento a irp

## Indicatori di rendimento 


#### TASSO DI RENDIMENTO NOMINALE

Indicatore di rendimento di un’obbligazione a tasso fisso pari al rapporto tra una singola cedola periodica e il valore nominale del titolo
TRN = (Cedola / Valore Nominale) x 100
- TRN è il Tasso di Rendimento Nominale (espresso in percentuale),
- Cedola è l'importo della singola cedola periodica,
- Valore Nominale è il valore nominale dell'obbligazione o del titolo.

In [28]:
def calcola_TRN(cedola, valore_nominale):

    trn = (cedola / valore_nominale) * 100
    return trn

# Esempio di utilizzo:
cedola = 40  # Importo della singola cedola periodica
valore_nominale = 1000  # Valore nominale dell'obbligazione o del titolo

tasso_rendimento_nominale = calcola_TRN(cedola, valore_nominale)
print("Tasso di Rendimento Nominale (TRN):", tasso_rendimento_nominale, "%")


Tasso di Rendimento Nominale (TRN): 4.0 %


#### TASSO DI RENDIMENTO IMMEDIATO

Rapporto tra i frutti periodici di un investimento e il capitale investito: TRI = (Cedola / Corso secco) x 100.
dove C = Cedole; CS = Corso secco (Prezzo di mercato di titoli con cedola).
È un indicatore parziale della redditività di un titolo, poiché considera solo il reddito derivante dallo stacco della cedola, in riferimento all’intervallo di tempo che intercorre tra lo stacco di due cedole successive.


In [29]:
def tasso_rendimento_immediato(somma_cedole, prezzo_acquisto):
    return somma_cedole / prezzo_acquisto

# Esempio di utilizzo
somma_cedole_annuali = 50
prezzo_acquisto_obbligazione = 950

rendimento_immediato = tasso_rendimento_immediato(somma_cedole_annuali, prezzo_acquisto_obbligazione)
print(f"Il tasso di rendimento immediato è: {rendimento_immediato:.2%}")


Il tasso di rendimento immediato è: 5.26%


#### TASSO DI RENDIMENTO EFFETTIVO A SCADENZA

Tasso di Rendimento Effettivo a Scadenza (TRES) è una misura più completa per quantificare la redditività di un investimento in titoli di debito, poiché tiene conto di molteplici caratteristiche dell'obbligazione e considera tutti gli aspetti che influenzano il rendimento effettivo dell'investimento.

La formula per calcolare il Tasso di Rendimento Effettivo a Scadenza è data da:

$TRES = \left[ \sum \frac{F_i}{(1 + TRES)^{t_i}} / VA \right]^{1/T} - 1$



Dove:
- `VA` rappresenta il valore attuale del titolo, ovvero il prezzo di acquisto dell'obbligazione (corso "tel-quel").
- `T` rappresenta il numero di periodi di detenzione del titolo fino alla scadenza.
- `F_i` rappresenta i flussi periodici corrisposti al possessore, come ad esempio le cedole annuali.
- `t_i` rappresenta il numero di periodi trascorsi dall'acquisto dell'obbligazione fino al ricevimento di ciascun flusso `F_i`.

Il TRES assume che il titolo venga detenuto fino alla scadenza e che le cedole ricevute siano reinvestite al tasso TRES stesso.


Il TRES supera i limiti insiti nel tasso di rendimento nominale e nel tasso di rendimento immediato, poiché considera tutte le caratteristiche dell'obbligazione. In sostanza, il tasso immediato considera solo il reddito staccato (cedole), mentre il TRES considera tutte le cedole fino a scadenza, i proventi ottenibili da loro reinvestimento allo stesso tasso e i profitti o le perdite da differenza tra valore di acquisto e di rimborso.

Il TRES è una misura più accurata per valutare la redditività di un investimento in obbligazioni perché tiene conto di tutte le componenti che influenzano il rendimento effettivo del titolo.


In [30]:
import numpy as np

In [31]:
import numpy as np
from scipy.optimize import newton

# Flussi di cassa dell'obbligazione
cashflows = np.array([10, 10, 110])

# Periodi di tempo (anni)
tempo = np.arange(1, 4)

# Prezzo corrente dell'obbligazione
prezzo_corrente_obbligazione = 102.531

def ytm_function(ytm, cashflows, tempo, prezzo_corrente_obbligazione):
    """
    Funzione che calcola il prezzo dell'obbligazione in base a un dato YTM.
    """
    return np.sum(cashflows / (1 + ytm) ** tempo) - prezzo_corrente_obbligazione

# Calcolo del YTM utilizzando il metodo di Newton-Raphson
ytm = newton(ytm_function, x0=0.05, args=(cashflows, tempo, prezzo_corrente_obbligazione))

print("Yield to Maturity (YTM)=TRES: {:.2%}".format(ytm))



Yield to Maturity (YTM)=TRES: 9.00%


## INDICATORI DI LIQUIDITA'


#### La Duration Nominale (DN)
è il periodo che intercorre tra il momento dell’emissione del titolo e la data di 
rimborso dell’ultima quota di capitale. DN = tcn.

In [32]:
# Duration Nominale (dato fornito o noto)
duration_nominale = 5  # Ad esempio, 5 anni

print("Duration Nominale (DN):", duration_nominale)


Duration Nominale (DN): 5


#### La Duration Residua (DR) 
è il periodo tra la sottoscrizione del titolo e la data di rimborso dell’ultima quota 
di capitale. Si differenzia dalla durata nominale perché il sottraendo non è la data dell’emissione, cioè 0, ma 
è la data di acquisto del titolo. Nella DN confrontiamo il momento dell’emissione con il momento 
dell’acquisto del titolo e la sua scadenza finale, mentre nella DR confrontiamo il momento dell’acquisto del 
titolo e la sua scadenza finale. DR = tcn – tK

In [33]:
def calcola_duration_residua(tc, tK):
    """
    Calcola la Duration Residua (DR) di un titolo.

    Parametri:
    tc (float): Termine alla Consegna, il numero di anni tra la data di emissione e la data di rimborso dell'ultima quota di capitale.
    tK (float): Numero di anni trascorsi tra la data di acquisto del titolo e la data di rimborso dell'ultima quota di capitale.

    Returns:
    float: La Duration Residua (DR) del titolo.
    """
    return tc - tK

# Esempio di utilizzo:
termine_consegna = 5      # Termine alla Consegna (in anni)
anni_trascorsi = 2        # Numero di anni trascorsi dall'acquisto alla scadenza
duration_residua = calcola_duration_residua(termine_consegna, anni_trascorsi)
print("Duration Residua (DR):", duration_residua)


Duration Residua (DR): 3


#### La Vita Media Probabile (VMP). 
È la media aritmetica ponderata dei periodi di scadenza dei rimborsi delle 
sole quote di capitale, ove i coefficienti di ponderazione sono il peso di ogni C (flusso di cassa) sul valore 
complessivo di Capitale rimborsato. Corrisponde alla sommatoria delle produttorie di ogni scadenza con il 
flusso di cassa a titolo di capitale maturato, normalizzata con la sommatoria dei flussi di cassa a titolo di 
capitale che vengono dati fino a scadenza

$$Vita Media Probabile (VMP) = \frac{\sum_{i=1}^{n} (t_i \cdot C_i)}{\sum_{i=1}^{n} C_i}$$


- Vita Media Probabile (VMP): Rappresenta la Vita Media Probabile dell'obbligazione o del portafoglio di obbligazioni.
- n: Indica il numero di flussi di cassa associati alle sole quote di capitale dell'obbligazione.
- t_i: Rappresenta il periodo di tempo residuo fino alla scadenza del capitale per il flusso di cassa C_i.
- C_i: Rappresenta l'importo del flusso di cassa a titolo di capitale (rimborso del capitale) al periodo i.

In altre parole, la VMP è calcolata come la somma dei prodotti tra i periodi residui dei flussi di cassa a titolo di capitale e i relativi importi dei flussi di cassa, il tutto diviso per la somma totale degli importi dei flussi di cassa a titolo di capitale.

La Vita Media Probabile fornisce una stima della durata media ponderata dei flussi di cassa a titolo di capitale, e viene spesso utilizzata per valutare la durata effettiva degli investimenti in obbligazioni, poiché tiene conto sia della durata dei flussi di cassa sia delle quantità dei flussi stessi.

In [34]:
def calcola_vita_media_probabile(flussi_capitale, periodi_scadenza):
    """
    Calcola la Vita Media Probabile (VMP) di un'obbligazione o di un portafoglio di obbligazioni.

    Parametri:
    flussi_capitale (list): Lista dei flussi di cassa a titolo di capitale (rimborso del capitale) in ordine di scadenza.
    periodi_scadenza (list): Lista dei periodi residui fino alla scadenza del capitale per ciascun flusso di cassa.

    Returns:
    float: La Vita Media Probabile (VMP) dell'obbligazione o del portafoglio di obbligazioni.
    """
    total_capitale = sum(flussi_capitale)
    weighted_sum = sum(t * C for t, C in zip(periodi_scadenza, flussi_capitale))
    return weighted_sum / total_capitale

# Esempio di utilizzo:
flussi_capitale = [50, 50, 50, 50, 1050]  # Lista dei flussi di cassa a titolo di capitale
periodi_scadenza = [1, 2, 3, 4, 5]        # Lista dei periodi residui fino alla scadenza del capitale
vita_media_probabile = calcola_vita_media_probabile(flussi_capitale, periodi_scadenza)
print("Vita Media Probabile (VMP):", vita_media_probabile)


Vita Media Probabile (VMP): 4.6


MP ≈ 4.6, indica la durata media ponderata dei flussi di cassa a titolo di capitale dell'obbligazione o del portafoglio di obbligazioni. In altre parole, è una stima del tempo medio in anni che ci si aspetta di mantenere l'investimento prima di ricevere l'ultimo rimborso di capitale.

Il valore VMP di circa 4.6 suggerisce che ci si aspetta di ricevere l'ultimo rimborso di capitale dell'obbligazione o del portafoglio di obbligazioni in circa 4.6 anni. Questo dato è ottenuto considerando i flussi di cassa a titolo di capitale e i relativi periodi residui fino alla scadenza.

### La Vita Media Matematica (VMM) 
è data dalla media ponderata dei periodi di scadenza dei rimborsi delle 
quote di capitale, ove i coefficienti di ponderazione sono il peso dei valori attuali di ogni C sul valore 
complessivo del Capitale rimborsato. La differenza rispetto alla VMP sta nel modo con cui vengono misurati 
i flussi di cassa. Nel VMP i flussi di cassa a titolo di capitale vengono sommati uno dopo l’altro, mentre nella 
VMM vengono sommati considerando il valore finanziario del tempo. Quindi, opera un trasferimento alla 
data di oggi, un’attualizzazione di tutte le serie di flussi di cassa a titolo di capitale che vengono generati dal 
titolo, per cui viene considerato per ogni flusso di cassa il suo fattore di sconto, attualizzato ad oggi per il 
suo orizzonte temporale.

$$Vita Media Matematica (VMM) = \frac{\sum_{i=1}^{n} t_i \cdot \frac{C_i}{(1 + r)^{t_i}}}{\sum_{i=1}^{n} \frac{C_i}{(1 + r)^{t_i}}}
$$


- Vita Media Matematica (VMM): Rappresenta la Vita Media Matematica dell'obbligazione o del portafoglio di obbligazioni.
- n: Indica il numero di flussi di cassa a titolo di capitale dell'obbligazione.
- t_i: Rappresenta il periodo di tempo residuo fino alla scadenza del capitale per il flusso di cassa C_i.
- C_i: Rappresenta l'importo del flusso di cassa a titolo di capitale (rimborso del capitale) al periodo i.
- r: Rappresenta il tasso di sconto, che è il tasso di rendimento richiesto o opportunità di investimento.

In [35]:
def calcola_vita_media_matematica(flussi_capitale, periodi_scadenza, tasso_di_sconto):
    """
    Calcola la Vita Media Matematica (VMM) di un'obbligazione o di un portafoglio di obbligazioni.

    Parametri:
    flussi_capitale (list): Lista dei flussi di cassa a titolo di capitale (rimborso del capitale) in ordine di scadenza.
    periodi_scadenza (list): Lista dei periodi residui fino alla scadenza del capitale per ciascun flusso di cassa.
    tasso_di_sconto (float): Il tasso di sconto utilizzato per attualizzare i flussi di cassa futuri.

    Returns:
    float: La Vita Media Matematica (VMM) dell'obbligazione o del portafoglio di obbligazioni.
    """
    attualizzati = [C / (1 + tasso_di_sconto) ** t for C, t in zip(flussi_capitale, periodi_scadenza)]
    total_attualizzati = sum(attualizzati)
    weighted_sum = sum(t * a for t, a in zip(periodi_scadenza, attualizzati))
    return weighted_sum / total_attualizzati

# Esempio di utilizzo:
flussi_capitale = [50, 50, 50, 50, 1050]  # Lista dei flussi di cassa a titolo di capitale (incluso il rimborso nominale)
periodi_scadenza = [1, 2, 3, 4, 5]       # Lista dei periodi residui fino alla scadenza del capitale
tasso_di_sconto = 0.05                   # Tasso di sconto del 5% (0.05 in valore decimale)
vita_media_matematica = calcola_vita_media_matematica(flussi_capitale, periodi_scadenza, tasso_di_sconto)
print("Vita Media Matematica (VMM):", vita_media_matematica)


Vita Media Matematica (VMM): 4.545950504162361


Il valore 4.54 ottenuto come risultato dalla Vita Media Matematica (VMM) indica la durata media ponderata dei flussi di cassa a titolo di capitale dell'obbligazione o del portafoglio di obbligazioni, considerando l'attualizzazione dei flussi futuri utilizzando il tasso di sconto specificato (nel nostro esempio, il tasso di sconto è 5%).

Più specificamente, il valore 4.54 rappresenta il tempo medio in anni che ci si aspetta di mantenere l'investimento prima di ricevere l'ultimo rimborso di capitale, considerando il valore finanziario del tempo e l'effetto dell'attualizzazione dei flussi di cassa futuri.

Questo valore può essere utile per gli investitori per valutare la durata media ponderata dell'investimento e per comprendere quanto tempo ci si aspetta di trattenere l'obbligazione o il portafoglio di obbligazioni prima di ricevere l'ultimo rimborso di capitale.

#### La Durata Media Ponderata (DMP) 
è data dalla media artimetica ponderata dei periodi di scadenza dei 
flussi finanziari futuri (interessi + capitale), ove i coefficienti di ponderazione sono il peso di ogni flusso sul 
valore complessivo dei flussi. 
$$DMP = (c1 * f1 + c2 * f2 + ... + cn * fn) / (f1 + f2 + ... + fn)$$
Dove:

- DMP è la Durata Media Ponderata.
- c1, c2, ..., cn rappresentano i periodi di scadenza dei flussi finanziari 1, 2, ..., n.
- f1, f2, ..., fn rappresentano i valori dei flussi finanziari 1, 2, ..., n.


In [36]:
def calcola_dmp(periodi_scadenza, valori_flussi):
    if len(periodi_scadenza) != len(valori_flussi):
        raise ValueError("Le liste devono avere la stessa lunghezza.")

    numeratore = sum([periodo * valore for periodo, valore in zip(periodi_scadenza, valori_flussi)])
    denominatore = sum(valori_flussi)

    dmp = numeratore / denominatore
    return dmp

# Dati forniti:
flussi_capitale = [50, 50, 50, 50, 1050]  # Lista dei flussi di cassa a titolo di capitale (incluso il rimborso nominale)
periodi_scadenza = [1, 2, 3, 4, 5]        # Lista dei periodi di scadenza

# Calcolo della DMP:
dmp = calcola_dmp(periodi_scadenza, flussi_capitale)
print("La Durata Media Ponderata (DMP) è:", dmp)


La Durata Media Ponderata (DMP) è: 4.6


In generale, la DMP è un indicatore di sensibilità di un investimento o di un titolo di debito ai cambiamenti dei tassi di interesse. Più alta è la DMP, maggiore sarà la sensibilità del titolo agli eventuali cambiamenti dei tassi di interesse.

Una DMP di 4.6,  suggerisce che i flussi finanziari futuri (interessi e capitale) sono distribuiti in media su un periodo di circa 4.6 unità di tempo ponderate dai valori dei flussi. 

#### La durata media finanziaria (o duration)
di un'obbligazione è definita come scadenza media dei flussi di cassa attesi, ponderata per il contributo del valore attuale di ciascun flusso alla formazione del prezzo.
A un'elevata duration corrisponde un'elevata sensibilità del prezzo del titolo al variare del tasso di rendimento e viceversa. Per tale motivo la duration viene utilizzata quale indicatore di rischio dei titoli obbligazionari. Essa permette di misurare (con una certa approssimazione) la variazione del prezzo di un titolo obbligazionario a seguito della variazione del livello dei tassi di interesse.
La duration è pari alla durata anagrafica per i titoli privi di cedola (zero coupon bond), mentre è sempre inferiore alla durata nel caso di titoli muniti di cedola. A parità di scadenza e di altre condizioni, la duration è più elevata per i titoli con cedola relativamente bassa e tende a diminuire all'aumentare del tasso di rendimento.

$$
MacD = \sum_{t=1}^{n} \left( \frac{1}{(1+y)^{t}} \times t \times CF_t \right) / PV
$$
Dove:

- MacD rappresenta la Durata di Macaulay dell'obbligazione.
- y è il rendimento alla scadenza (yield to maturity) dell'obbligazione.
- tf è il tempo in anni fino a quando viene ricevuto il flusso di cassa (pagamento di cedola o valore di rimborso).
- CFf rappresenta l'importo del flusso di cassa per il rispettivo periodo.
- PV è il valore attuale di tutti i flussi di cassa

In [37]:
# Dati forniti
flussi_capitale = [50, 50, 50, 50, 1050]  # Lista dei flussi di cassa a titolo di capitale (incluso il rimborso nominale)
periodi_scadenza = [1, 2, 3, 4, 5]       # Lista dei periodi di scadenza corrispondenti ai flussi di cassa
ytm = 0.05                              # Rendimento alla scadenza (Yield to Maturity) del 5%

# Calcolo del valore attuale dei flussi di cassa
def valore_attuale(flusso, periodo, ytm):
    return flusso / (1 + ytm)**periodo

# Calcolo della Durata di Macaulay
def durata_macaulay(flussi, periodi, ytm):
    presente_valore = [valore_attuale(flusso, periodo, ytm) for flusso, periodo in zip(flussi, periodi)]
    pv_totale = sum(presente_valore)
    
    durata = sum([(valore * periodo) for valore, periodo in zip(presente_valore, periodi)]) / pv_totale
    return durata

# Calcolo della Durata di Macaulay utilizzando i dati forniti
durata = durata_macaulay(flussi_capitale, periodi_scadenza, ytm)
print(f"La Durata di Macaulay dell'obbligazione è: {durata:.2f} anni")


La Durata di Macaulay dell'obbligazione è: 4.55 anni


La Durata di Macaulay indica quanto tempo mediamente ci vorrà per recuperare il costo dell'obbligazione attraverso i flussi di cassa futuri, tenendo conto del tasso di interesse corrente. In generale, una Durata di Macaulay maggiore indica una maggiore sensibilità del prezzo dell'obbligazione ai cambiamenti dei tassi di interesse.

Quindi, nel caso specifico, una Durata di Macaulay di 4.55 anni indica che ci vorranno circa 4.55 anni per recuperare l'investimento in quella specifica obbligazione considerando i flussi di cassa attesi e il rendimento alla scadenza del 5%.

#### Duration modificata

La Duration Modificata è una misura finanziaria utilizzata per stimare la variazione percentuale del prezzo di un'obbligazione rispetto alle variazioni dei tassi di rendimento effettivo a scadenza (TRES). Essa rappresenta un'approssimazione lineare della variazione del prezzo in risposta ai cambiamenti dei tassi di interesse.

La formula per calcolare la Duration Modificata è la seguente:


$$Duration Modificata = MacD / (1 + TRES)$$

Dove:

MacD rappresenta la Durata di Macaulay dell'obbligazione.
TRES indica il tasso di rendimento effettivo a scadenza, calcolato come rapporto tra il prezzo attuale dell'obbligazione e il valore attuale di tutti i flussi di cassa.
La Duration Modificata è espressa in anni ed è un'indicazione della percentuale di variazione del prezzo dell'obbligazione per una variazione di un punto percentuale del tasso di rendimento effettivo a scadenza.

In [38]:
# Dati forniti
flussi_capitale = [50, 50, 50, 50, 1050]  # Lista dei flussi di cassa a titolo di capitale (incluso il rimborso nominale)
periodi_scadenza = [1, 2, 3, 4, 5]       # Lista dei periodi di scadenza corrispondenti ai flussi di cassa
ytm = 0.05                               # Rendimento alla scadenza (Yield to Maturity) del 5%

# Calcolo del valore attuale dei flussi di cassa
def valore_attuale(flusso, periodo, ytm):
    return flusso / (1 + ytm)**periodo

# Calcolo della Durata di Macaulay
def durata_macaulay(flussi, periodi, ytm):
    presente_valore = [valore_attuale(flusso, periodo, ytm) for flusso, periodo in zip(flussi, periodi)]
    pv_totale = sum(presente_valore)
    
    durata = sum([(valore * periodo) for valore, periodo in zip(presente_valore, periodi)]) / pv_totale
    return durata

# Calcolo della Duration Modificata
def duration_modificata(macaulay_duration, ytm):
    return macaulay_duration / (1 + ytm)

# Calcolo della Durata di Macaulay utilizzando i dati forniti
durata_macaulay = durata_macaulay(flussi_capitale, periodi_scadenza, ytm)

# Calcolo della Duration Modificata utilizzando la Durata di Macaulay
duration_modificata = duration_modificata(durata_macaulay, ytm)

print(f"Durata di Macaulay dell'obbligazione: {durata_macaulay:.2f} anni")
print(f"Duration Modificata dell'obbligazione: {duration_modificata:.2f} anni")


Durata di Macaulay dell'obbligazione: 4.55 anni
Duration Modificata dell'obbligazione: 4.33 anni


Se la Duration Modificata è 4.33 anni, significa che ci aspettiamo che il prezzo dell'obbligazione varierà di circa il 4.33% per ogni variazione di un punto percentuale (1%) nei tassi di rendimento effettivo a scadenza (YTM).

Quindi, se i tassi di interesse aumentano di 1 punto percentuale (ad esempio, dal 5% al 6%), possiamo calcolare l'effetto sulla variazione percentuale del prezzo dell'obbligazione utilizzando la Duration Modificata:

Variazione percentuale del prezzo dell'obbligazione = - Duration Modificata × Variazione percentuale dei tassi di interesse

Variazione percentuale del prezzo dell'obbligazione = - 4.33% × 1% = -0.0433

Quindi, con un aumento dei tassi di interesse del 1%, ci aspettiamo che il prezzo dell'obbligazione diminuisca di circa il 0.0433%, ovvero dello 0.0433 punti percentuali.

In [39]:
duration_modificata = 4.33               # Duration Modificata

# Calcolo dell'effetto di un aumento del 1% dei tassi di interesse
variazione_tassi_aumento = 0.01
variazione_prezzo_aumento = -duration_modificata * variazione_tassi_aumento

# Calcolo dell'effetto di una diminuzione del 1% dei tassi di interesse
variazione_tassi_diminuzione = -0.01
variazione_prezzo_diminuzione = -duration_modificata * variazione_tassi_diminuzione

print(f"Effetto di un aumento del 1% dei tassi di interesse sul prezzo dell'obbligazione: {variazione_prezzo_aumento:.2f}%")
print(f"Effetto di una diminuzione del 1% dei tassi di interesse sul prezzo dell'obbligazione: {variazione_prezzo_diminuzione:.2f}%")

Effetto di un aumento del 1% dei tassi di interesse sul prezzo dell'obbligazione: -0.04%
Effetto di una diminuzione del 1% dei tassi di interesse sul prezzo dell'obbligazione: 0.04%


#### convessità
$$Convexity=Cx=\frac{1}{B_0}\frac{1}{((1+r)^2}\sum_{t=1}^{T}(t+t^2)\frac{C_t}{(1+r)^t}$$



In [40]:
# Dati forniti
flussi_capitale = [50, 50, 50, 50, 1050]
periodi_scadenza = [1, 2, 3, 4, 5]
ytm = 0.05

# Calcolo del prezzo attuale dell'obbligazione (B0)
present_value = [flusso / (1 + ytm)**periodo for flusso, periodo in zip(flussi_capitale, periodi_scadenza)]
B0 = sum(present_value)

# Calcolo della Convessità
convessita = sum([(t + t**2) * flusso / (B0 * (1 + ytm)**t) for t, flusso in zip(periodi_scadenza, flussi_capitale)])

print(f"Convessità dell'obbligazione con 5 periodi: {convessita:.2f}")


Convessità dell'obbligazione con 5 periodi: 26.39


Quando il valore della Convessità è troppo elevato, può indicare che l'obbligazione ha una maggiore sensibilità ai cambiamenti dei tassi di interesse, il che potrebbe comportare un maggiore rischio di prezzo in determinate situazioni di mercato. Un'elevata Convessità può portare a due principali considerazioni:

Maggiore incertezza: Un'obbligazione con alta Convessità è più sensibile ai cambiamenti dei tassi di interesse, il che potrebbe aumentare l'incertezza riguardo alla direzione futura del prezzo dell'obbligazione. Questo può essere particolarmente problematico in periodi di volatilità dei tassi di interesse o incertezza economica.

Maggiore esposizione al rischio: Un'alta Convessità può significare che l'obbligazione è più suscettibile a fluttuazioni di prezzo in risposta ai cambiamenti dei tassi di interesse. Questo potrebbe comportare un maggiore rischio di prezzo, soprattutto se i tassi di interesse subiscono movimenti significativi.